atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresDdlTable.java

```diff
@@ -118,6 +123,48 @@ public void checkDatabaseVersion() {
 
     @Override
     public void compactInternally() {
+        if (compactionSemaphore.tryAcquire()) {
+            try {
+                if (shouldRunCompaction()) {
+                    runCompactOnTable();
+                }
+            } finally {
+                compactionSemaphore.release();
+            }
+        }
+    }
+
+    @VisibleForTesting
+    boolean shouldRunCompaction() {
+        long compactIntervalMillis = config.compactInterval().toMilliseconds();
+        return compactIntervalMillis <= 0 || getMillisSinceLastCompact() >= compactIntervalMillis;
+    }
+
+    /**
+     * Returns the number of milliseconds since the last compaction, or Long.MAX_VALUE if
+     * compaction has never run.
+     */
+    private long getMillisSinceLastCompact() {
+        AgnosticResultSet rs = conns.get().selectResultSetUnregisteredQuery(
+                "SELECT FLOOR(EXTRACT(EPOCH FROM GREATEST( "
+                        + "  last_vacuum, last_autovacuum, last_analyze, last_autoanalyze"
+                        + "))*1000) AS last, "
+                        + "FLOOR(EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000) AS current "
+                        + "FROM pg_stat_user_tables WHERE relname = ?",
```

Comment:
nice. I don't know if you can push some of the math on the DB side, because NULL - numbers = NULL, you could maybe do something like

```sql
SELECT COALESCE(
  FLOOR(EXTRACT(EPOCH FROM CURRENT_TIMESTAMP)*1000) - FLOOR(EXTRACT(EPOCH FROM GREATEST(a, b, c, d))*1000), CAST(POWER(2, 63) - 1 AS BIGINT)) FROM pg_stat_user_tables WHERE relname = ?
```

but anyway the key thing is to get this down to 1 DB query which we've done, and it may be a better idea to keep the logic in better understood Java.

Reply:
Dont think we need too much perf optimization here. Sticking to the current values.